### PR TITLE
Move flight controller node to YAMSPy library

### PR DESCRIPTION
**Pull request by Tuxliri**
## Description

- Please include a summary of the change and which issue is fixed.
- Please also include relevant motivation and context.
- List any dependencies that are required for this change.
- Explicitly mention if the current changes can (or could) break any existing interface or functionality.
- If this pull request results in from a fix to a GitHub issue, give a link to it.

## Type of change

Please mark the relevant options with X:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change is constrained only to one package, any interfaces to nodes from other packages are maintained
- [ ] This change is constrained only to packages in this repository, any interfaces to nodes from other repositories are maintained
- [ ] This change required a documentation update
- [ ] This change is only a documentation update
- [ ] This change restructures the repository (at level higher than package level)
- [ ] This change requires a change in the Duckiebook (if yes, please describe and, if possible, make a pull request for an edit in the book and link it here)

## Tests

Please describe all tests ran and their outcome. It is highly recommended that you at least do the tests listed by default, even if you only changed the documentation.

- [ ] Made the Duckiebot move with keyboard control
- [ ] Visually checked the published camera stream
- [ ] Verified that the lane following demo works
- [ ] Verified that the indefinite navigation demo works

## Style compliance
Please make sure that all your changes conform to the Duckietown (code style)[link] and code (documentation style)[link] guidelines:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I checked that this pull request is compliant with the Duckietown code style
- [ ] I checked that this pull request is compliant with the Duckietown code documentation style

## Next steps
Please describe if you have any questions, comments, or doubts towards the Duckietown software guardians regarding implementation, code style, running tests, documentation, etc.

## FOR THE REVIEWERS
__IMPORTANT:__ DO NOT CHANGE THIS SECTION!

- [ ] Functionality (is the change necessary and is it implemented in the most appropriate way): @liampaull or @afdaniele
- [ ] Integration testing (verify the tests and run additional if necessary): @?
- [ ] Code style compliance: @gibernas
- [ ] Documentation compliance: @aleksandarpetrov

_To the reviewers:_ If you deem necessary, request code edits, more tests, documentation update, or additional reviewing. __Do not review your own code!__

_PR Template, last edit 23 Aug 2019_

### CODE REVIEW

File: packages/flight_controller_driver/src/flight_controller_node.py [hunk 1]
```diff
@@ -251,35 +266,41 @@ def _srv_calibrate_imu_cb(self, _):
...
+                        time.sleep(0.2)
+                        
+                        # Sometimes we receive a [0,0,0] SENSOR_DATA array, we need to skip this for
+                        # averaging
+                        if np.sum(self._board.SENSOR_DATA['accelerometer']) > 50:
+                            data["x"] += self._board.SENSOR_DATA['accelerometer'][0]
+                            data["y"] += self._board.SENSOR_DATA['accelerometer'][1]
+                            data["z"] += self._board.SENSOR_DATA['accelerometer'][2]
+
+                            counter += 1
```
Comment (Tuxliri) [inline packages/flight_controller_driver/src/flight_controller_node.py]: This is a bit of a hack, should we investigate it further?

File: packages/flight_controller_driver/src/flight_controller_node.py [hunk 2]
```diff
@@ -472,31 +501,65 @@ def _read_motor_pwm_signals(self):
         """
         try:
             # read m1, m2, m3, m4
-            self._board.getData(MultiWii.MOTOR)
+            if self._board.send_RAW_msg(MSPy.MSPCodes['MSP_MOTOR'], data=[]):
+                        dataHandler = self._board.receive_msg()
+                        self._board.process_recv_data(dataHandler)
```
Comment (afdaniele) [inline packages/flight_controller_driver/src/flight_controller_node.py]: wrong indentation here

File: packages/flight_controller_driver/src/flight_controller_node.py (hunk 1 above)
Comment (afdaniele) [reply packages/flight_controller_driver/src/flight_controller_node.py]: leave the note for now

File: packages/flight_controller_driver/src/flight_controller_node.py [hunk 3]
```diff
@@ -23,9 +26,22 @@
 
 from dt_class_utils import DTReminder
 from duckietown.dtros import DTROS, NodeType, DTParam, ParamType
-from h2r_multiwii import MultiWii
-from h2r_multiwii.types import MultiWiiRpyPid
+from yamspy import MSPy
+
+
+@dataclass
+class MultiWiiRpyPid:
```
Comment (afdaniele) [inline packages/flight_controller_driver/src/flight_controller_node.py]: If we are dropping "MultiWii" let's not name anything after it

File: packages/flight_controller_driver/src/flight_controller_node.py [hunk 4]
```diff
@@ -629,8 +696,8 @@ def on_shutdown(self):
         if self._board is None:
             return
         self.loginfo("Disarming!")
-        self._board.sendCMD(8, MultiWii.SET_RAW_RC, self.rc_command(DroneMode.DISARMED))
-        self._board.receiveDataPacket()
+        self._command = self.rc_command(DroneMode.ARMED)
+        self._switch_to_mode(DroneMode.ARMED, quiet=True)
```
Comment (afdaniele) [inline packages/flight_controller_driver/src/flight_controller_node.py]: ARMED or DISARMED?

File: packages/flight_controller_driver/src/flight_controller_node.py [hunk 5]
```diff
@@ -251,25 +266,31 @@ def _srv_calibrate_imu_cb(self, _):
...
-                        self._board.getData(MultiWii.RAW_IMU)
-                        data["x"] += self._board.rawIMU["ax"]
-                        data["y"] += self._board.rawIMU["ay"]
-                        data["z"] += self._board.rawIMU["az"]
-                        counter += 1
-                        time.sleep(1.0 / self._frequency["imu"])
-                    except Exception:
-                        pass
+                        self._board.fast_read_imu()
+                        time.sleep(0.2)
```
Comment (afdaniele) [inline packages/flight_controller_driver/src/flight_controller_node.py:281]: can we do this more often? `20 * 200ms = 4s`

File: packages/flight_controller_driver/src/flight_controller_node.py [hunk 6]
```diff
@@ -251,25 +266,31 @@ def _srv_calibrate_imu_cb(self, _):
...
+                        
+                        # Sometimes we receive a [0,0,0] SENSOR_DATA array, we need to skip this for
+                        # averaging
+                        if np.sum(self._board.SENSOR_DATA['accelerometer']) > 50:
```
Comment (afdaniele) [inline packages/flight_controller_driver/src/flight_controller_node.py]: don't we want `np.sum(np.abs(...))` instead? accelerations can be negative

**Review by afdaniele — CHANGES_REQUESTED** [review]

File: packages/flight_controller_driver/src/flight_controller_node.py (hunk 5 above)
Comment (Tuxliri) [reply packages/flight_controller_driver/src/flight_controller_node.py:281]: It seems to drop more packages if done more often.

File: packages/flight_controller_driver/src/flight_controller_node.py (hunk 4 above)
Comment (Tuxliri) [reply packages/flight_controller_driver/src/flight_controller_node.py]: Definitely DISARMED

File: packages/flight_controller_driver/src/flight_controller_node.py (hunk 5 above)
Comment (afdaniele) [reply packages/flight_controller_driver/src/flight_controller_node.py:281]: define "drop packages"

Comment (Tuxliri) [reply packages/flight_controller_driver/src/flight_controller_node.py:281]: I meant read more almost-zero acceleration vectors rather than drop packages.

Comment (Tuxliri) [reply packages/flight_controller_driver/src/flight_controller_node.py:281]: We can remove this as it duplicates the purpose of `MSP_ACC_CALIBRATION`

**Pull request merged**